# Auditoría de Calidad de Datos - NovaMarket
**Archivo:** `S03_PD_Martinez_Sandoval_Torres_Olmo_AuditoriaCalidad.ipynb`

**Dataset:** `NovaMarket_datos_crudos_2.csv`

**Entrega grupal**

**Equipo:**
- Diego Alejandro Sandoval
- Santiago Torres
- Juan Andrés Martínez
- Laura Valentina Olmos

Este cuaderno audita el dataset crudo de NovaMarket en las **6 dimensiones de calidad de datos**: completitud, exactitud, consistencia, validez, unicidad y oportunidad. 
Cada hallazgo se cuantifica con conteos exactos, y al final se resume en un catálogo de problemas priorizado por severidad (`S03_PD_Martinez_Sandoval_Torres_Olmo_CatalogoProblemas.csv`).

In [34]:
import pandas as pd
df = pd.read_csv('NovaMarket_datos_crudos 2.csv')
print(df.shape)
df.head()

(640000, 20)


,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
0,P505195,C45650,15/11/2025,Tienda,Efectivo,bogota,1100103,Moda,Chaqueta impermeable,28700.0,1,28700.0,32,cliente405195@novamarket.com.co,Bajo,Platino,El domiciliario fue muy amable,14.9,2026-04-17,0
1,P677836,C84260,2026-03-05,Web,Efectivo,MZL,1700101,Electrónica,Parlante bluetooth,195700.0,2,391400.0,73,cliente577836@yahoo.es,Medio,Bronce,El producto llegó dañado,12.9,2026-06-10,0
2,P387869,C164166,2025-03-13,Tienda,PayPal,Cúcuta,5400101,Juguetería,Rompecabezas 1000 piezas,597600.0,-3,-1792800.0,32,cliente287869@novamarket.com.co,Bajo,Oro,El domiciliario fue muy amable,14.3,2026-02-26,0
3,P481069,C98406,27/06/2025,App,transferencia,Cucuta,5400103,deportes,Colchoneta de yoga,322900.0,1,322900.0,45,cliente381069@gmail.com,Medio,Platino,Rápido y sin problemas,9.5,2026-05-14,0
4,P247920,C163018,01/05/2025,Web,Efectivo,Bucaramanga,6800112,Hogar,Juego de sábanas,222400.0,1,222400.0,60,cliente147920@novamarket.com.co,Alto,Plata,NaN,12.3,2026-07-13,0


In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 640000 entries, 0 to 639999
Data columns (total 20 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   id_pedido                  640000 non-null  object 
 1   id_cliente                 640000 non-null  object 
 2   fecha_pedido               640000 non-null  object 
 3   canal_compra               640000 non-null  object 
 4   metodo_pago                640000 non-null  object 
 5   ciudad_tienda              610072 non-null  object 
 6   codigo_postal              640000 non-null  int64  
 7   categoria_producto         640000 non-null  object 
 8   producto                   640000 non-null  object 
 9   precio_unitario            618360 non-null  float64
 10  unidades_vendidas          640000 non-null  int64  
 11  monto_compra               618360 non-null  float64
 12  edad_cliente               640000 non-null  int64  
 13  correo_cliente             53

---
## 1. Auditoría de Completitud

Revisamos cuántos valores faltantes (`NaN`) tiene cada columna, en cantidad y en porcentaje sobre el total de **640.000** registros.

In [36]:
df.isnull().sum()

id_pedido                         0
id_cliente                        0
fecha_pedido                      0
canal_compra                      0
metodo_pago                       0
ciudad_tienda                 29928
codigo_postal                     0
categoria_producto                0
producto                          0
precio_unitario               21640
unidades_vendidas                 0
monto_compra                  21640
edad_cliente                      0
correo_cliente               105312
nivel_satisfaccion            46469
nivel_lealtad                     0
comentario_cliente            90806
peso_pedido_kg                    0
fecha_actualizacion_stock         0
fuga_cliente                      0
dtype: int64

In [37]:
(df.isnull().mean() * 100).round(2)

id_pedido                     0.00
id_cliente                    0.00
fecha_pedido                  0.00
canal_compra                  0.00
metodo_pago                   0.00
ciudad_tienda                 4.68
codigo_postal                 0.00
categoria_producto            0.00
producto                      0.00
precio_unitario               3.38
unidades_vendidas             0.00
monto_compra                  3.38
edad_cliente                  0.00
correo_cliente               16.46
nivel_satisfaccion            7.26
nivel_lealtad                 0.00
comentario_cliente           14.19
peso_pedido_kg                0.00
fecha_actualizacion_stock     0.00
fuga_cliente                  0.00
dtype: float64

In [38]:
# Revisar si hay "falsos completos" en columnas categóricas clave (valores como "N/A" o "-")
df["nivel_lealtad"].value_counts(dropna=False)

nivel_lealtad
Bronce     160420
Oro        160222
Plata      159912
Platino    159446
Name: count, dtype: int64

In [39]:
df["nivel_satisfaccion"].value_counts(dropna=False)

nivel_satisfaccion
Alto      166739
Medio     165777
Bajo      165309
NaN        46469
alto       12803
medio      12661
bajo       12585
BAJO        6582
 Alto       6554
 Medio      6444
Medio       6429
 Bajo       6383
MEDIO       6347
Alto        6335
ALTO        6315
Bajo        6268
Name: count, dtype: int64

## **Hallazgos de completitud**

En el dataset de **640.000 registros**, se identificaron **6 columnas** con valores faltantes y **14 columnas** con el 100% de registros completos:

- `correo_cliente`: **105.312** valores faltantes (**16.46%**). Es el campo más incompleto: sin correo no se puede contactar directamente a 1 de cada 6 clientes.
- `comentario_cliente`: **90.806** valores faltantes (**14.19%**).
- `nivel_satisfaccion`: **46.469** valores faltantes (**7.26%**).
- `ciudad_tienda`: **29.928** valores faltantes (**4.68%**).
- `precio_unitario`: **21.640** valores faltantes (**3.38%**). Campo financiero clave: sin precio no se puede calcular el ingreso de esos pedidos.
- `monto_compra`: **21.640** valores faltantes (**3.38%**) — son exactamente los mismos registros que tienen `precio_unitario` nulo, porque `monto_compra` se calcula a partir de ese campo.
- Al revisar `nivel_lealtad` y `nivel_satisfaccion` con `value_counts()`, no se encontraron "falsos completos" (no aparecen valores como "N/A", "-" o "Sin dato"); los faltantes son `NaN` reales.

---
## 2. Auditoría de Exactitud

Aplicamos reglas de dominio sobre `precio_unitario`, `unidades_vendidas`, `edad_cliente`, `peso_pedido_kg` y `monto_compra` para encontrar valores que, aunque tengan el tipo de dato correcto, son imposibles en el contexto del negocio.

In [40]:
# Regla 1: precio unitario negativo
precio_negativo = df[df["precio_unitario"] < 0]
print("Precio unitario negativo:", len(precio_negativo))

# Regla 2: precio unitario absurdamente alto
precio_absurdo = df[df["precio_unitario"] > 10_000_000]
print("Precio unitario > $10.000.000:", len(precio_absurdo))

Precio unitario negativo: 19629
Precio unitario > $10.000.000: 6164


In [41]:
# Regla 3: unidades vendidas negativas
unidades_negativas = df[df["unidades_vendidas"] < 0]
print("Unidades vendidas negativas:", len(unidades_negativas))

# Regla 4: unidades vendidas en cero
unidades_cero = df[df["unidades_vendidas"] == 0]
print("Unidades vendidas en cero:", len(unidades_cero))

# Regla 5: unidades vendidas absurdas (más de 100 en un solo pedido)
unidades_absurdas = df[df["unidades_vendidas"] > 100]
print("Unidades vendidas > 100:", len(unidades_absurdas))

Unidades vendidas negativas: 18550
Unidades vendidas en cero: 9327
Unidades vendidas > 100: 18526


In [42]:
# Regla 6: edad del cliente fuera de rango [0, 120]
edad_negativa = df[df["edad_cliente"] < 0]
edad_absurda = df[df["edad_cliente"] > 120]
print("Edad negativa:", len(edad_negativa))
print("Edad > 120 años:", len(edad_absurda))
print("Total edad fuera de rango:", len(edad_negativa) + len(edad_absurda))

Edad negativa: 9254
Edad > 120 años: 18575
Total edad fuera de rango: 27829


In [43]:
# Regla 7 (propia del equipo): peso del pedido no puede ser negativo
peso_negativo = df[df["peso_pedido_kg"] < 0]
print("Peso de pedido negativo:", len(peso_negativo), "-> sin problemas detectados en esta regla")

Peso de pedido negativo: 0 -> sin problemas detectados en esta regla


In [44]:
# Regla 8 (propia del equipo): monto_compra debe ser igual a precio_unitario x unidades_vendidas
esperado = df["precio_unitario"] * df["unidades_vendidas"]
comparables = df["precio_unitario"].notna() & df["monto_compra"].notna()
monto_inconsistente = comparables & ((df["monto_compra"] - esperado).abs() > 1)
print("Registros comparables:", comparables.sum())
print("monto_compra inconsistente con precio_unitario x unidades_vendidas:", monto_inconsistente.sum(), "-> sin problemas detectados en esta regla")

Registros comparables: 618360
monto_compra inconsistente con precio_unitario x unidades_vendidas: 0 -> sin problemas detectados en esta regla


## **Hallazgos de exactitud**
(8 reglas de dominio aplicadas, incluyendo 2 reglas propias del equipo)

- `precio_unitario`: **19.629** registros con precio negativo (**3.07%**) — imposible en una venta, distorsiona cualquier suma de ingresos.
- `precio_unitario`: **6.164** registros con precio mayor a $10.000.000 (**0.96%**).
- `unidades_vendidas`: **18.550** registros con cantidad negativa (**2.90%**).
- `unidades_vendidas`: **9.327** registros con cantidad en cero (**1.46%**) — un pedido sin unidades no debería existir.
- `unidades_vendidas`: **18.526** registros con más de 100 unidades en un solo pedido (**2.89%**) — con un máximo de 9.999, claro error de captura.
- `edad_cliente`: **27.829** registros con edad fuera del rango 0-120 años (**4.35%**): 9.254 negativas y 18.575 mayores a 120.
- **Regla propia del equipo:** `peso_pedido_kg` nunca es negativo — **0** registros afectados, sin problemas detectados en este campo.
- **Regla propia del equipo:** `monto_compra` siempre coincide exactamente con `precio_unitario x unidades_vendidas` en los **618.360** registros donde ambos campos están presentes — **0** inconsistencias. Confirma que `monto_compra` es un campo calculado y no una fuente independiente de error.

---
## 3. Auditoría de Consistencia

Con `value_counts()` revisamos si el mismo valor real está escrito de formas distintas (mayúsculas, tildes, abreviaturas, espacios) en cuatro columnas categóricas, y sumamos manualmente los registros de las variantes que representan el mismo valor real.

In [45]:
df["canal_compra"].value_counts(dropna=False)

canal_compra
Marketplace     121426
Web             121377
Tienda          121269
App             121147
tienda           15684
web              15535
app              15477
marketplace      15332
APP               7824
Marketplace       7787
 Marketplace      7780
 Tienda           7754
 Web              7748
TIENDA            7741
WEB               7737
App               7729
MARKETPLACE       7710
Web               7709
Tienda            7648
 App              7586
Name: count, dtype: int64

Agrupando manualmente las variantes que representan el mismo canal (sumando mayúsculas, minúsculas y espacios extra):

- **Marketplace** = **160.035** registros
- **Web** = **160.106** registros
- **Tienda** = **160.096** registros
- **App** = **159.763** registros

`canal_compra` tiene **20 valores de texto distintos** que representan solo **4 canales reales**.

In [46]:
df["metodo_pago"].value_counts(dropna=False)

metodo_pago
Transferencia     127962
Tarjeta           127631
PayPal            127583
Efectivo          127452
paypal             13145
efectivo           12955
transferencia      12932
tarjeta            12807
 Efectivo           6593
Tarjeta             6581
 Transferencia      6544
Transferencia       6529
 Tarjeta            6514
EFECTIVO            6513
PAYPAL              6416
Efectivo            6406
PayPal              6393
TARJETA             6385
 PayPal             6375
TRANSFERENCIA       6284
Name: count, dtype: int64

Agrupando manualmente las variantes que representan el mismo método de pago:

- **Transferencia** = **160.251** registros
- **Tarjeta** = **159.918** registros
- **PayPal** = **159.912** registros
- **Efectivo** = **159.919** registros

`metodo_pago` tiene **20 valores de texto distintos** que representan solo **4 métodos reales**.

In [47]:
df["categoria_producto"].value_counts(dropna=False)

categoria_producto
Belleza         56378
Electrónica     56261
Deportes        56019
Hogar           55846
Juguetería      55499
Moda            55300
moda            20354
hogar           20352
deportes        20351
belleza         20273
electronica     10433
JUGUETERÍA      10332
HOGAR           10290
Electrónica     10267
DEPORTES        10264
Deportes        10214
Belleza         10195
Moda            10182
 Juguetería     10170
Hogar           10157
 Deportes       10147
electrónica     10142
 Moda           10136
juguetería      10133
BELLEZA         10114
ELECTRÓNICA     10085
 Belleza        10052
Juguetería      10047
jugueteria      10029
MODA            10025
 Electrónica    10008
 Hogar           9945
Name: count, dtype: int64

Agrupando manualmente las variantes que representan la misma categoría:

- **Belleza** (`Belleza` + `belleza` + `Belleza ` + `BELLEZA` + ` Belleza`) = **107.012** registros
- **Electrónica** (`Electrónica` + `electronica` + `Electrónica ` + `electrónica` + `ELECTRÓNICA` + ` Electrónica`) = **107.196** registros
- **Deportes** (`Deportes` + `deportes` + `DEPORTES` + `Deportes ` + ` Deportes`) = **106.995** registros
- **Hogar** (`Hogar` + `hogar` + `HOGAR` + `Hogar ` + ` Hogar`) = **106.590** registros
- **Juguetería** (`Juguetería` + `JUGUETERÍA` + ` Juguetería` + `juguetería` + `Juguetería ` + `jugueteria`) = **106.210** registros
- **Moda** (`Moda` + `moda` + `Moda ` + ` Moda` + `MODA`) = **105.997** registros

`categoria_producto` tiene **32 valores de texto distintos** que representan solo **6 categorías reales**.

In [48]:
df["ciudad_tienda"].value_counts(dropna=False)

ciudad_tienda
NaN              29928
Villavicencio    27634
Pereira          27549
Cali             25462
Bucaramanga      25461
                 ...  
 Bogotá           2662
bogota d.c.       2654
BOGOTÁ            2646
BOGOTA DC         2638
BOG               2630
Name: count, Length: 89, dtype: int64

Agrupando manualmente las variantes que representan la misma ciudad (mayúsculas, tildes, espacios y abreviaturas como `BOG`, `MDE`, `CTG`, `BAQ`, `BGA`, `MZL`, `SMR`):

- **Bogotá** = **50.810** registros
- **Barranquilla** = **50.884** registros
- **Cali** = **51.002** registros
- **Santa Marta** = **50.755** registros
- **Medellín** = **50.775** registros
- **Cartagena** = **51.079** registros
- **Bucaramanga** = **50.840** registros
- **Manizales** = **50.991** registros
- **Ibagué** = **50.636** registros
- **Cúcuta** = **50.735** registros
- **Villavicencio** = **50.702** registros
- **Pereira** = **50.863** registros
- Faltantes (`NaN`) = **29.928** registros (ya contados en la sección de completitud)

`ciudad_tienda` tiene **88 valores de texto distintos** (sin contar `NaN`) que representan solo **12 ciudades reales**.

## **Hallazgos de consistencia**

- `canal_compra`: **20** valores distintos representan solo **4** canales reales.
- `metodo_pago`: **20** valores distintos representan solo **4** métodos reales.
- `categoria_producto`: **32** valores distintos representan solo **6** categorías reales.
- `ciudad_tienda`: **88** valores distintos representan solo **12** ciudades reales — es la columna con más variantes de escritura de todo el dataset.
- En las cuatro columnas, el problema es el mismo: mayúsculas/minúsculas, tildes, espacios extra y abreviaturas hacen que `value_counts()` y cualquier `groupby()` traten un mismo valor real como si fueran valores distintos, lo que rompe cualquier reporte agregado por canal, método de pago, categoría o ciudad hasta que se normalice el texto.

---
## 4. Auditoría de Validez

Revisamos si el correo tiene el símbolo `@` y si la fecha de pedido se puede convertir con `pd.to_datetime(errors="coerce")`.

In [49]:
# Correos sin @ (solo entre los que sí tienen valor, para no mezclar con los faltantes de la sección 1)
correo_no_nulo = df["correo_cliente"].notna()
correos_invalidos = df[correo_no_nulo & ~df["correo_cliente"].str.contains("@", na=False)]

print("Correos no nulos:", correo_no_nulo.sum())
print("Correos sin @:", len(correos_invalidos))

Correos no nulos: 534688
Correos sin @: 51532


In [50]:
# Conversión de fecha_pedido
fecha_pedido_conv = pd.to_datetime(df["fecha_pedido"], errors="coerce")

fechas_invalidas = df[fecha_pedido_conv.isna() & df["fecha_pedido"].notna()]
print("Fechas de pedido que no se pudieron convertir:", len(fechas_invalidas))

Fechas de pedido que no se pudieron convertir: 223053


C:\Users\DIEGO\AppData\Local\Temp\ipykernel_20772\4193334140.py:2: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  fecha_pedido_conv = pd.to_datetime(df["fecha_pedido"], errors="coerce")


In [51]:
# Fechas de pedido futuras (posteriores a hoy)
hoy = pd.Timestamp.today()
fechas_futuras = df[fecha_pedido_conv > hoy]

print("Fecha de referencia (hoy):", hoy.date())
print("Fechas de pedido futuras:", len(fechas_futuras))

Fecha de referencia (hoy): 2026-08-05
Fechas de pedido futuras: 16485


## **Hallazgos de validez**

- `correo_cliente`: de los **534.688** correos no nulos, **51.532** (**9.64%** de los no nulos, **8.05%** del total) no contienen el símbolo `@`.
- `fecha_pedido`: al convertir con `pd.to_datetime(errors="coerce")`, **223.053** registros (**34.85%**) no se pudieron convertir — el dataset mezcla formatos de fecha distintos (`AAAA-MM-DD` y `D/M/AAAA`) en la misma columna, y el método simple asume un solo formato para toda la columna, lo que confirma otro ángulo del problema de **consistencia** de formato de fechas.
- `fecha_pedido`: de las fechas que sí se lograron convertir, **16.485** (**2.58%**) quedan como posteriores a hoy — probablemente hay más fechas futuras "escondidas" entre las 223.053 no convertidas, así que este número es un piso, no el total real.

---
## 5. Auditoría de Unicidad

Buscamos filas 100% duplicadas y duplicados en `id_pedido`, que debería ser una clave única, con `duplicated()`.

In [52]:
# Filas completamente duplicadas
print("Filas 100% duplicadas:", df.duplicated().sum())

Filas 100% duplicadas: 20000


In [53]:
# Duplicados por id_pedido (clave que debería ser única)
duplicados_id = df[df.duplicated(subset=["id_pedido"], keep=False)].sort_values("id_pedido")

print("Filas involucradas en algún id_pedido duplicado:", len(duplicados_id))
duplicados_id[["id_pedido", "id_cliente", "fecha_pedido", "producto"]].head(6)

Filas involucradas en algún id_pedido duplicado: 40000


,id_pedido,id_cliente,fecha_pedido,producto
172693,P100005,C164871,2025-01-28,Rompecabezas 1000 piezas
174646,P100005,C164871,2025-01-28,Rompecabezas 1000 piezas
418817,P100010,C76626,17/8/2025,Olla a presión
368632,P100010,C76626,17/8/2025,Olla a presión
168620,P100023,C46545,00/00/0000,Secador de pelo
103928,P100023,C46545,00/00/0000,Secador de pelo


## **Hallazgos de unicidad**

- **20.000** filas están 100% duplicadas: todas sus columnas son idénticas a otra fila (**3.13%**).
- `id_pedido` tiene **20.000** valores repetidos, lo que involucra **40.000** filas en total (**6.25%**). Al revisar manualmente algunas de esas filas, confirmamos que corresponden al mismo pedido reingresado dos veces.

---
## 6. Auditoría de Oportunidad

Calculamos los días transcurridos desde `fecha_actualizacion_stock` hasta hoy, y definimos un umbral de negocio para considerar el stock "desactualizado".

In [54]:
fecha_actualizacion_conv = pd.to_datetime(df["fecha_actualizacion_stock"], errors="coerce")
dias_desde_actualizacion = (hoy - fecha_actualizacion_conv).dt.days

dias_desde_actualizacion.describe(percentiles=[.5, .75, .85, .9])

count    640000.000000
mean        161.928248
std         289.660190
min        -510.000000
50%          94.000000
75%         143.000000
85%         162.000000
90%         605.000000
max        1312.000000
Name: fecha_actualizacion_stock, dtype: float64

**¿Por qué un umbral de 180 días?** El percentil 85 de "días desde la última actualización" es de 162 días, y el siguiente valor de referencia (percentil 90) salta directo a 605 días (más de un año y medio). Ese salto tan grande entre el percentil 85 (162 días) y el percentil 90 (605 días) muestra que hay dos grupos bien separados: la mayoría del inventario se actualiza en menos de 6 meses, y un grupo aparte lleva mucho más tiempo sin tocarse. Por eso 180 días cae justo en la "zona segura" entre ambos grupos, y no es un número arbitrario.

In [55]:
umbral_dias = 180
stock_desactualizado = df[dias_desde_actualizacion > umbral_dias]

print(f"Registros con stock sin actualizar hace más de {umbral_dias} días:", len(stock_desactualizado))

Registros con stock sin actualizar hace más de 180 días: 66069


In [56]:
# Fechas de actualización de stock en el futuro (imposible)
stock_fecha_futura = df[fecha_actualizacion_conv > hoy]

print("Registros con fecha_actualizacion_stock en el futuro:", len(stock_fecha_futura))

Registros con fecha_actualizacion_stock en el futuro: 21722


## **Hallazgos de oportunidad**

- **66.069** registros (**10.32%**) llevan más de **180 días** sin actualizar su stock.
- **21.722** registros (**3.39%**) tienen `fecha_actualizacion_stock` en el futuro, lo cual es imposible: el stock no puede haberse actualizado en una fecha que todavía no llega.

---
## 7. Resumen ejecutivo y catálogo de problemas

Consolidamos todos los hallazgos anteriores en el catálogo `S03_PD_Martinez_Sandoval_Torres_Olmo_CatalogoProblemas.csv`, ordenado de mayor a menor severidad.

**Criterio de severidad** (combina dos factores, no es solo el porcentaje):
- **Crítico**: el error hace que el dato sea lógica o financieramente imposible (ej. precio negativo), o afecta a más del 15% del dataset.
- **Alto**: afecta entre el 5% y el 15% del dataset, o un porcentaje menor pero en un campo financiero/operativo clave (precio, unidades, id_pedido).
- **Medio**: afecta entre 1% y 5% del dataset en un campo no financiero, o son variantes de texto que degradan el reporting pero no rompen cálculos.
- **Bajo**: afecta a menos del 1% del dataset y tiene bajo impacto operativo.

In [57]:
catalogo = [
    ("Validez", "fecha_pedido", "Fecha que no se pudo convertir con pd.to_datetime (mezcla de formatos + fechas inexistentes)", 223053, 34.85, "Crítico",
     "Afecta a más de un tercio del dataset y compromete cualquier análisis temporal o de ingresos por periodo."),
    ("Completitud", "correo_cliente", "Valor faltante (NaN)", 105312, 16.46, "Crítico",
     "Bloquea la comunicación directa con 1 de cada 6 clientes; supera el umbral de 15%."),
    ("Exactitud", "precio_unitario", "Precio negativo", 19629, 3.07, "Crítico",
     "Campo financiero directo; un precio negativo es lógicamente imposible."),
    ("Completitud", "comentario_cliente", "Valor faltante (NaN)", 90806, 14.19, "Alto",
     "Limita el análisis cualitativo de retroalimentación de clientes."),
    ("Consistencia", "ciudad_tienda", "Variantes del mismo valor (mayúsculas, tildes, abreviaturas): 88 valores para 12 ciudades reales", 314864, 49.20, "Alto",
     "Sin normalizar, cualquier segmentación geográfica o logística queda rota; es la columna con más variantes de todo el dataset."),
    ("Consistencia", "categoria_producto", "Variantes del mismo valor: 32 valores para 6 categorías reales", 304697, 47.61, "Alto",
     "Rompe el reporting de ventas por categoría, una de las vistas más usadas del negocio."),
    ("Consistencia", "canal_compra", "Variantes del mismo valor: 20 valores para 4 canales reales", 154781, 24.18, "Alto",
     "Sin normalizar no se puede reportar ventas por canal de forma confiable."),
    ("Consistencia", "metodo_pago", "Variantes del mismo valor: 20 valores para 4 métodos reales", 129372, 20.21, "Alto",
     "Sin normalizar no se puede reportar ventas por método de pago de forma confiable."),
    ("Unicidad", "id_pedido", "Pedido duplicado (filas involucradas)", 40000, 6.25, "Alto",
     "Cada pedido duplicado infla artificialmente el conteo de ventas e ingresos en reportes agregados."),
    ("Exactitud", "unidades_vendidas", "Cantidad negativa", 18550, 2.90, "Alto",
     "Campo que multiplica directamente el ingreso; un valor negativo es financieramente imposible."),
    ("Exactitud", "unidades_vendidas", "Cantidad absurda (más de 100 en un solo pedido, hasta 9.999)", 18526, 2.89, "Alto",
     "Cada outlier distorsiona fuertemente cualquier suma de ingresos o inventario."),
    ("Oportunidad", "fecha_actualizacion_stock", "Stock sin actualizar hace más de 180 días", 66069, 10.32, "Alto",
     "Decisiones de inventario tomadas sobre datos de más de medio año de antigüedad."),
    ("Completitud", "nivel_satisfaccion", "Valor faltante (NaN)", 46469, 7.26, "Alto",
     "Limita el análisis de satisfacción de cliente para ese subconjunto."),
    ("Exactitud", "edad_cliente", "Edad imposible (negativa o mayor a 120 años)", 27829, 4.35, "Medio",
     "Afecta la segmentación demográfica, sin impacto financiero directo."),
    ("Completitud", "ciudad_tienda", "Valor faltante (NaN)", 29928, 4.68, "Medio",
     "Limita el análisis geográfico y logístico para ese subconjunto de pedidos."),
    ("Validez", "correo_cliente", "Formato inválido (sin @)", 51532, 8.05, "Medio",
     "Correos no utilizables para contacto; volumen menor que el problema de completitud del mismo campo."),
    ("Completitud", "precio_unitario", "Valor faltante (NaN)", 21640, 3.38, "Alto",
     "Campo financiero clave: sin precio no se puede calcular el ingreso de esos pedidos."),
    ("Exactitud", "precio_unitario", "Precio absurdamente alto (> $10.000.000)", 6164, 0.96, "Medio",
     "Volumen bajo; revisar caso a caso, posible error de dígitos de más."),
    ("Validez", "fecha_pedido", "Fecha de pedido posterior a hoy", 16485, 2.58, "Medio",
     "Una venta no puede haber ocurrido en el futuro; es un piso, hay más casos ocultos entre las fechas no convertidas."),
    ("Oportunidad", "fecha_actualizacion_stock", "Fecha de actualización en el futuro", 21722, 3.39, "Medio",
     "El stock no puede haberse actualizado en una fecha que todavía no llega."),
    ("Exactitud", "unidades_vendidas", "Cantidad en cero", 9327, 1.46, "Medio",
     "Un pedido sin unidades no debería existir como venta; revisar si son cancelaciones mal registradas."),
    ("Exactitud", "peso_pedido_kg", "Peso negativo (regla de dominio propia)", 0, 0.00, "Bajo",
     "Sin problemas detectados; se documenta como validación limpia."),
    ("Exactitud", "monto_compra", "Inconsistente con precio_unitario x unidades_vendidas (regla de dominio propia)", 0, 0.00, "Bajo",
     "Sin problemas detectados; confirma que monto_compra es un campo calculado y confiable."),
]

catalogo_df = pd.DataFrame(
    catalogo,
    columns=["dimension", "columna", "problema", "registros_afectados", "pct_afectado", "severidad", "justificacion"]
)

orden_severidad = {"Crítico": 0, "Alto": 1, "Medio": 2, "Bajo": 3}
catalogo_df["orden_severidad"] = catalogo_df["severidad"].map(orden_severidad)
catalogo_df = catalogo_df.sort_values(["orden_severidad", "pct_afectado"], ascending=[True, False]).drop(columns="orden_severidad").reset_index(drop=True)

catalogo_df

,dimension,columna,problema,registros_afectados,pct_afectado,severidad,justificacion
0,Validez,fecha_pedido,Fecha que no se pudo convertir con pd.to_datet...,223053,34.85,Crítico,Afecta a más de un tercio del dataset y compro...
1,Completitud,correo_cliente,Valor faltante (NaN),105312,16.46,Crítico,Bloquea la comunicación directa con 1 de cada ...
2,Exactitud,precio_unitario,Precio negativo,19629,3.07,Crítico,Campo financiero directo; un precio negativo e...
3,Consistencia,ciudad_tienda,"Variantes del mismo valor (mayúsculas, tildes,...",314864,49.20,Alto,"Sin normalizar, cualquier segmentación geográf..."
4,Consistencia,categoria_producto,Variantes del mismo valor: 32 valores para 6 c...,304697,47.61,Alto,"Rompe el reporting de ventas por categoría, un..."
5,Consistencia,canal_compra,Variantes del mismo valor: 20 valores para 4 c...,154781,24.18,Alto,Sin normalizar no se puede reportar ventas por...
6,Consistencia,metodo_pago,Variantes del mismo valor: 20 valores para 4 m...,129372,20.21,Alto,Sin normalizar no se puede reportar ventas por...
7,Completitud,comentario_cliente,Valor faltante (NaN),90806,14.19,Alto,Limita el análisis cualitativo de retroaliment...
8,Oportunidad,fecha_actualizacion_stock,Stock sin actualizar hace más de 180 días,66069,10.32,Alto,Decisiones de inventario tomadas sobre datos d...
9,Completitud,nivel_satisfaccion,Valor faltante (NaN),46469,7.26,Alto,Limita el análisis de satisfacción de cliente ...


In [58]:
catalogo_df.to_csv("S03_PD_Martinez_Sandoval_Torres_Olmo_CatalogoProblemas.csv", index=False, encoding="utf-8-sig")
print("Catálogo guardado:", len(catalogo_df), "filas")
catalogo_df["severidad"].value_counts()

Catálogo guardado: 23 filas


severidad
Alto       11
Medio       7
Crítico     3
Bajo        2
Name: count, dtype: int64

**Conclusión general:** el dataset crudo de NovaMarket (640.000 registros) tiene problemas en las 6 dimensiones evaluadas. Los tres hallazgos de severidad **crítica** son: la mezcla de formatos en `fecha_pedido` (afecta al 34.85% del dataset), los correos faltantes (16.46%) y los precios negativos (financieramente imposibles). Los problemas de **consistencia** son los de mayor volumen en términos relativos: `canal_compra` y `metodo_pago` afectan al 100% de las filas si se cuentan todas sus variantes de escritura, y `ciudad_tienda` tiene 88 formas distintas de escribir solo 12 ciudades reales. Todos se solucionan con un proceso de normalización de texto y agrupación manual relativamente simple. El catálogo completo con los 23 hallazgos, su severidad y justificación queda disponible en `S03_PD_Martinez_Sandoval_Torres_Olmo_CatalogoProblemas.csv`.